File: app/src/main/java/com/bigjelly/temporun/MainActivity.java
package com.bigjelly.temporun;

import android.app.Activity;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.util.Log;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.SeekBar;
import android.widget.TextView;

import com.bigjelly.temporun.player.AdvancedAudioPlayer;
import com.bigjelly.temporun.player.Mp3AudioTrackPlayer;

import java.util.Locale;

import kotlin.Unit;


public class MainActivity extends Activity {
    private static final String TAG = "MainActivity";

    private Mp3AudioTrackPlayer player = new Mp3AudioTrackPlayer();

    private AdvancedAudioPlayer advancedPlayer = new AdvancedAudioPlayer();
    private boolean isUserTrackingSeekBar = false; // 标记用户是否正在拖拽进度条

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Button start = findViewById(R.id.start);
        start.setOnClickListener(v -> {
            player.playFromAssets(this, "music.mp3");
        });
        Button stop = findViewById(R.id.stop);
        stop.setOnClickListener(v -> {
            player.stop();
        });

        SeekBar seekBar = findViewById(R.id.seek_bar);
        TextView tvCurrent = findViewById(R.id.tv_current_time);
        TextView tvTotal = findViewById(R.id.tv_total_time);
        CheckBox cbLoop = findViewById(R.id.cb_loop);
        Button btnPlay = findViewById(R.id.btn_play);
        Button btnPause = findViewById(R.id.btn_pause);
        Button btnResume = findViewById(R.id.btn_resume);
        Button btnStop = findViewById(R.id.btn_stop);

        advancedPlayer.setOnProgressListener((currentMs, totalMs) -> {
            if (!isUserTrackingSeekBar) {
                Log.d(TAG, "onProgressListener: currentMs: " + currentMs + ", totalMs: " + totalMs);
                seekBar.setMax(totalMs.intValue());
                seekBar.setProgress(currentMs.intValue());
                tvCurrent.setText(formatTime(currentMs));
                tvTotal.setText(formatTime(totalMs));
            }
            return Unit.INSTANCE;
        });
        advancedPlayer.setOnCompletionListener(() -> {
            tvCurrent.setText("00:00");
            seekBar.setProgress(0);
            return Unit.INSTANCE;
        });

        seekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {

            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                if (fromUser) {
                    tvCurrent.setText(formatTime((long) progress));
                }
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {
                isUserTrackingSeekBar = true;
            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {
                isUserTrackingSeekBar = false;
                advancedPlayer.seekTo(seekBar.getProgress());
            }
        });

        btnPlay.setOnClickListener(view -> advancedPlayer.playFromAssets(MainActivity.this, "music.mp3", cbLoop.isChecked()));
        btnPause.setOnClickListener(view -> advancedPlayer.pause());
        btnStop.setOnClickListener(view -> {
            advancedPlayer.stop();
            tvCurrent.setText("00:00");
            seekBar.setProgress(0);
        });
        btnResume.setOnClickListener(view -> advancedPlayer.resume());
        cbLoop.setOnCheckedChangeListener((buttonView, isChecked) -> advancedPlayer.setLooping(isChecked));

    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        player.stop();
        advancedPlayer.stop();
    }

    private String formatTime(Long millis) {
        long totalSeconds = millis / 1000;
        long minutes = totalSeconds / 60;
        long seconds = totalSeconds % 60;
        return String.format(Locale.getDefault(), "%02d:%02d", minutes, seconds);
    }
}

File: app/src/main/java/com/bigjelly/temporun/MainActivity.kt
package com.bigjelly.temporun

import android.app.Activity
import android.os.Bundle
import android.widget.Button
import android.widget.CheckBox
import android.widget.SeekBar
import android.widget.TextView
import com.bigjelly.temporun.player.AdvancedAudioPlayer
import com.bigjelly.temporun.player.Mp3AudioTrackPlayer
import java.util.Locale

class MainActivity : Activity() {

    private val player = Mp3AudioTrackPlayer()
    private val advancedPlayer = AdvancedAudioPlayer()
    private var isUserTrackingSeekBar = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val start = findViewById<Button>(R.id.start)
        start.setOnClickListener {
            player.playFromAssets(this, "music.mp3")
        }

        val stop = findViewById<Button>(R.id.stop)
        stop.setOnClickListener {
            player.stop()
        }

        val seekBar = findViewById<SeekBar>(R.id.seek_bar)
        val tvCurrent = findViewById<TextView>(R.id.tv_current_time)
        val tvTotal = findViewById<TextView>(R.id.tv_total_time)
        val cbLoop = findViewById<CheckBox>(R.id.cb_loop)
        val btnPlay = findViewById<Button>(R.id.btn_play)
        val btnPause = findViewById<Button>(R.id.btn_pause)
        val btnResume = findViewById<Button>(R.id.btn_resume)
        val btnStop = findViewById<Button>(R.id.btn_stop)

        advancedPlayer.onProgressListener = { currentMs, totalMs ->
            if (!isUserTrackingSeekBar) {
                tvCurrent.text = formatTime(currentMs)
                tvTotal.text = formatTime(totalMs)
                if (totalMs > 0) {
                    seekBar.max = totalMs.toInt()
                    seekBar.progress = currentMs.toInt()
                }
            }
        }

        advancedPlayer.onCompletionListener = {
            tvCurrent.text = "00:00"
            seekBar.progress = 0
        }

        seekBar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(
                seekBar: SeekBar,
                progress: Int,
                fromUser: Boolean
            ) {
                if (fromUser) {
                    tvCurrent.text = formatTime(progress.toLong())
                }
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {
                isUserTrackingSeekBar = true
            }

            override fun onStopTrackingTouch(seekBar: SeekBar) {
                isUserTrackingSeekBar = false
                advancedPlayer.seekTo(seekBar.progress.toLong())
            }
        })

        btnPlay.setOnClickListener {
            advancedPlayer.playFromAssets(this, "music.mp3", cbLoop.isChecked)
        }
        btnPause.setOnClickListener { advancedPlayer.pause() }
        btnResume.setOnClickListener { advancedPlayer.resume() }
        btnStop.setOnClickListener {
            advancedPlayer.stop()
            tvCurrent.text = "00:00"
            seekBar.progress = 0
        }
        cbLoop.setOnCheckedChangeListener { _, isChecked ->
            advancedPlayer.setLooping(isChecked)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        player.stop()
        advancedPlayer.stop()
    }

    private fun formatTime(millis: Long): String {
        val totalSeconds = millis / 1000
        val minutes = totalSeconds / 60
        val seconds = totalSeconds % 60
        return String.format(Locale.getDefault(), "%02d:%02d", minutes, seconds)
    }
}
